cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -0,0 +1,212 @@
+                                                             GSON.toJson(tetherRequest));
+      if (response.getResponseCode() != 200) {
+        LOG.error("Failed to initiate tether with peer {}, response body: {}, code: {}",
+                  peerInfo.getName(), response.getResponseBody(), response.getResponseCode());
+      }
+    } catch (URISyntaxException | IOException e) {
+      LOG.error("Failed to send tether request to peer {}, endpoint {}",
+                peerInfo.getName(), peerInfo.getEndpoint());
+    }
+  }
+
+  private void initializeMessageIds() {
+    lastMessageIds = new HashMap<>();
+    List<String> topics;
+    try {
+      topics = store.getPeers().stream()
+        .map(PeerInfo::getName)
+        .collect(Collectors.toList());
+    } catch (IOException e) {
+      LOG.warn("Failed to get peer information", e);
+      return;
+    }
+
+    TransactionRunners.run(transactionRunner, context -> {
+      for (String topic : topics) {
+        String messageId = AppMetadataStore.create(context).retrieveSubscriberState(topic, SUBSCRIBER);
+        lastMessageIds.put(topic, messageId);
+      }
+    });
+  }
+
+
+  private void handleForbidden(PeerInfo peerInfo) throws IOException {
+    if (peerInfo.getTetheringStatus() != TetheringStatus.PENDING) {
+      LOG.debug("Ignoring tethering rejection message from {}, current state: {}", peerInfo.getName(),
+                peerInfo.getTetheringStatus());
+      return;
+    }
+    // Set tethering status to rejected.
+    store.updatePeerStatusAndTimestamp(peerInfo.getName(), TetheringStatus.REJECTED);
+  }
+
+  private void handleResponse(HttpResponse resp, PeerInfo peerInfo) throws IOException {
+    if (peerInfo.getTetheringStatus() == TetheringStatus.PENDING) {
+      LOG.debug("Peer {} transitioned to ACCEPTED state", peerInfo.getName());
+      store.updatePeerStatusAndTimestamp(peerInfo.getName(), TetheringStatus.ACCEPTED);
+    } else {
+      // Update last connection timestamp.
+      store.updatePeerTimestamp(peerInfo.getName());
+    }
+    processTetherControlResponse(resp.getResponseBodyAsString(StandardCharsets.UTF_8), peerInfo);
+  }
+
+  private void processTetherControlResponse(String message, PeerInfo peerInfo) {
+    TetheringControlResponse tetheringControlResponse = GSON.fromJson(message, TetheringControlResponse.class);
+    for (TetheringControlMessage tetheringControlMessage : tetheringControlResponse.getControlMessages()) {
+      switch (tetheringControlMessage.getType()) {
+        case KEEPALIVE:
+          LOG.trace("Got keeplive from {}", peerInfo.getName());
+          break;
+        case RUN_PIPELINE:
+          // TODO: add processing logic
+          break;
+      }
+    }
+
+    String lastMessageId = tetheringControlResponse.getLastMessageId();
+    lastMessageIds.put(peerInfo.getName(), lastMessageId);
+    TransactionRunners.run(transactionRunner, context -> {
```

Comment:
I think we can move the update of all messageIds at the end of the one iteration, instead of doing it per peer.

Reply:
done